File: haskell/Polyomino/Point.hs
module Point where

data Point = Point { x :: Int, y :: Int } deriving (Show, Read, Eq, Ord)

move :: Int -> Int -> Point -> Point
move dx dy (Point x y) = Point (x + dx) (y + dy)

rotateRight :: Point -> Point -> Point
rotateRight (Point x1 y1) (Point x2 y2) = Point newX newY
    where 
        newX = -(y2 - y1) + x1
        newY =  (x2 - x1) + y1

rotateLeft :: Point -> Point -> Point
rotateLeft (Point x1 y1) (Point x2 y2) = Point newX newY
    where 
        newX = (y2 - y1) + x1
        newY = -(x2 - x1) + y1

reflectVertically :: Int -> Point -> Point
reflectVertically rx (Point x y) = Point (2 * rx - x) y

reflectHorizontally :: Int -> Point -> Point
reflectHorizontally ry (Point x y) = Point x (2 * ry - y)

File: python/polyomino/point.py

class Point(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def move(self, dx, dy):
        return Point(self.x + dx, self.y + dy)

    def rotate_right(self, p):
        new_x = -(self.y - p.y) + p.x
        new_y = (self.x - p.x) + p.y
        return Point(new_x, new_y)

    def rotate_left(self, p):
        new_x = (self.y - p.y) + p.x
        new_y = -(self.x - p.x) + p.y
        return Point(new_x, new_y)

    def reflect_vertically(self, x):
        return Point(2 * x - self.x, self.y)

    def reflect_horizontally(self, y):
        return Point(self.x, 2 * y - self.y)

    def __hash__(self):
        return hash((self.x, self.y))

    def __eq__(self, p):
        return (self.x, self.y) == (p.x, p.y)

    def __cmp__(self, p):
        return cmp((self.x, self.y), (p.x, p.y))

    def __repr__(self):
        return "Point(" + str(self.x) + ", " + str(self.y) + ")"

